File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/assert-doctor-ran.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
# 🚧 GATE — refuse to proceed on an unverified / broken environment.
#
# The mandatory Step-0 environment check (doctor.sh / doctor.ps1) writes a
# machine-readable doctor.json fingerprint. This turns "run the doctor first"
# from prose into a real gate: if the doctor never ran, or ran and FAILED, the
# pipeline stops here with the exact remediation — instead of the agent
# improvising around a missing runtime, which is the #1 source of cross-user
# inconsistency at multi-user events.
#
# doctor.json is looked up in this order:
#   1. <workdir>/doctor.json         (when --workdir is given)
#   2. ~/.sigma-migration/doctor.json (the stable location doctor always writes)
#
# Usage:
#   ruby scripts/assert-doctor-ran.rb [--workdir DIR]
#     [--skip-doctor-gate REASON]   # waive — REQUIRED reason; name it in your report
#
# Exit codes:
#   0  doctor.json present and pass:true (or waived)
#   1  doctor.json missing, unreadable, or pass:false
#   2  usage error
require 'json'
require 'optparse'

opts = {}
OptionParser.new do |p|
  p.on('--workdir DIR') { |v| opts[:dir] = v }
  p.on('--tableau DIR', 'alias of --workdir') { |v| opts[:dir] = v }
  p.on('--skip-doctor-gate REASON',
       'waive the environment gate — REQUIRED reason string; name it in your report') do |v|
    opts[:skip] = v
  end
end.parse!

if opts[:skip]
  puts "[SKIP] environment gate WAIVED (#{opts[:skip]}) — name this in your report."
  exit 0
end

home_doctor = File.expand_path('~/.sigma-migration/doctor.json')
candidates = []
candidates << File.join(opts[:dir], 'doctor.json') if opts[:dir]
candidates << home_doctor
path = candidates.find { |p| File.exist?(p) }

def remediate
  warn '       Run the environment doctor FIRST, then re-run:'
  warn '         macOS/Linux/Git-Bash:  bash scripts/doctor.sh'
  warn '         Windows PowerShell:    powershell -ExecutionPolicy Bypass -File scripts\\doctor.ps1'
  warn '       Escape hatch (name it in your report): --skip-doctor-gate "<reason>".'
end

unless path
  warn '[FAIL] environment gate — no doctor.json found (the Step-0 environment check never ran).'
  remediate
  exit 1
end

begin
  d = JSON.parse(File.read(path))
rescue JSON::ParserError => e
  warn "[FAIL] environment gate — doctor.json at #{path} is unreadable: #{e.message}"
  remediate
  exit 1
end

rt = d['runtimes'] || {}
env_desc = "os=#{d['os']} shell=#{d['shell']} sandbox=#{d['sandbox_hint']} " \
           "runtimes=[#{rt.select { |_, v| v }.keys.join(',')}]"

if d['pass']
  puts "[PASS] environment gate — doctor.json OK (#{env_desc}). Source: #{path}"
  exit 0
end

warn "[FAIL] environment gate — the environment doctor reported blocking failures (#{env_desc}):"
Array(d['failures']).each { |f| warn "         ✗ #{f}" }
remediate
exit 1

File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/assert-doctor-ran.py
#!/usr/bin/env python3
"""Fail-closed Python gate for doctor.json and bootstrap.json."""

from __future__ import annotations

import argparse
import datetime as dt
import json
import os
import sys
from pathlib import Path


def warn(message: str) -> None:
    print(message, file=sys.stderr)


def remediate(profile: str | None = None) -> None:
    suffix = f" --runtime-profile {profile}" if profile else ""
    warn("       Run the ONE bootstrap command FIRST, then retry:")
    warn(f"         macOS/Linux/Git-Bash: bash scripts/bootstrap.sh{suffix}")
    warn(
        "         Windows PowerShell: powershell -ExecutionPolicy Bypass "
        f"-File scripts\\bootstrap.ps1{(' -RuntimeProfile ' + profile) if profile else ''}"
    )


def load_json(path: Path) -> dict:
    with path.open(encoding="utf-8-sig") as handle:
        value = json.load(handle)
    if not isinstance(value, dict):
        raise ValueError("top-level value must be an object")
    return value


def append_waiver(workdir: Path, reason: str) -> None:
    workdir.mkdir(parents=True, exist_ok=True)
    record = {
        "kind": "skip-flag-waived",
        "reason": reason,
        "detail": "--skip-doctor-gate",
        "recorded_at": dt.datetime.now(dt.timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
    }
    with (workdir / "offramps.jsonl").open("a", encoding="utf-8") as handle:
        handle.write(json.dumps(record, sort_keys=True) + "\n")


def first_existing(paths: list[Path]) -> Path | None:
    return next((path for path in paths if path.is_file()), None)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--workdir", "--tableau", dest="workdir")
    parser.add_argument("--runtime-profile", choices=("ruby", "python"))
    parser.add_argument("--skip-doctor-gate")
    args = parser.parse_args()

    workdir = Path(args.workdir).expanduser().resolve() if args.workdir else None
    if args.skip_doctor_gate:
        print(
            f"[SKIP] environment gate WAIVED ({args.skip_doctor_gate}) "
            "— name this in your report."
        )
        if workdir:
            append_waiver(workdir, args.skip_doctor_gate)
        return 0

    doctor_candidates = []
    if workdir:
        doctor_candidates.append(workdir / "doctor.json")
    doctor_candidates.append(Path.home() / ".sigma-migration" / "doctor.json")
    doctor_path = first_existing(doctor_candidates)
    if doctor_path is None:
        warn("[FAIL] environment gate — no doctor.json found.")
        remediate(args.runtime_profile)
        return 1

    try:
        doctor = load_json(doctor_path)
    except (OSError, ValueError, json.JSONDecodeError) as exc:
        warn(f"[FAIL] environment gate — doctor.json at {doctor_path} is unreadable: {exc}")
        remediate(args.runtime_profile)
        return 1

    threshold = int(os.environ.get("SIGMA_MAX_BEHIND", "50"))
    behind = doctor.get("behind_count")
    if isinstance(behind, int) and behind > threshold:
        warn(
            f"[FAIL] environment gate — skill is {behind} commit(s) behind "
            f"origin/main (> {threshold})."
        )
        return 1

    profile = doctor.get("runtime_profile") or {}
    selected = profile.get("selected")
    if args.runtime_profile and selected != args.runtime_profile:
        warn(
            "[FAIL] environment gate — doctor runtime profile mismatch: "
            f"expected {args.runtime_profile!r}, selected {selected!r}."
        )
        remediate(args.runtime_profile)
        return 1

    if not doctor.get("pass"):
        warn("[FAIL] environment gate — the doctor reported blocking failures:")
        for failure in doctor.get("failures") or []:
            warn(f"         ✗ {failure}")
        remediate(args.runtime_profile)
        return 1

    sentinel_candidates = []
    if workdir:
        sentinel_candidates.append(workdir / "bootstrap.json")
    sentinel_candidates.append(Path.home() / ".sigma-migration" / "bootstrap.json")
    sentinel_path = first_existing(sentinel_candidates)
    if sentinel_path is None:
        warn("[FAIL] environment gate — doctor passes but bootstrap.json is missing.")
        remediate(args.runtime_profile)
        return 1
    try:
        sentinel = load_json(sentinel_path)
    except (OSError, ValueError, json.JSONDecodeError) as exc:
        warn(f"[FAIL] environment gate — bootstrap.json at {sentinel_path} is unreadable: {exc}")
        return 1
    if not sentinel.get("doctor_pass"):
        warn("[FAIL] environment gate — bootstrap sentinel records doctor_pass=false.")
        return 1
    sentinel_profile = (sentinel.get("runtime_profile") or {}).get("selected")
    if selected and sentinel_profile != selected:
        warn(
            "[FAIL] environment gate — doctor/bootstrap runtime profiles differ: "
            f"{selected!r} vs {sentinel_profile!r}."
        )
        return 1

    runtimes = doctor.get("runtimes") or {}
    present = ",".join(name for name, available in runtimes.items() if available)
    print(
        f"[PASS] environment gate — profile={selected or 'legacy'} "
        f"runtimes=[{present}]; bootstrap sentinel {sentinel_path}. "
        f"Source: {doctor_path}"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
